src/python/dxpy/bindings/apollo/expression_matrix_transformation.py

```diff
@@ -0,0 +1,63 @@
+from copy import deepcopy
+
+def expression_transform(results_list):
+    """
+    results_list: list of dictionaries, each of the form
+    {
+        "feature_id":<feature_id>,
+        "sample_id":<sample_id>,
+        "expression":<expression>
+    }
+    Additional key:value pairs are possible, but are not included in the transformed output
+
+    How it works:
+    First, create a dictionary where each key is a sample id, and the values are the feature_id:expression pairs
+    Then, convert it to a list of dictionaries (the format that output handling is expecting)
+    Each item in the list is a dictionary containing one "sample_id":sample_id pair, and
+    multiple feature_id:expression pairs
+
+    Returns:
+    A list of dictionaries, each of the form:
+    {
+        "sample_id":<sample_id>,
+        <feature_id_1>:<expression_1>
+        <feature_id_2>:<expression_2>
+        ...
+        <feature_id_n>:<expression_n>
+    }
+    """
+    transformed_dict = {}
+    # create a dict of the form
+    # {
+    #   <sample_id>:{
+    #       <feature_id_1>:<expression_1>,
+    #       <feature_id_2>:<expression_2>,
+    #        etc.
+    #   }
+    # 
+    # }
+    for entry in results_list:
```

Comment:
How about starting with a `set` to store all unique `feature_id`s?

Reply:
That way there would be no need to keep track of `colnames` since all dicts will have the same set of `feature_id`s in the same, exact order. We would just fill in `None` whenever there is no matching `value` for the corresponding sample_id/feature_id pair.

In this case `write_expression_output` will also work without any tweaks (because all dicts will have the same keys in the same order).

Reply:
For the first point, I can use set in the first loop and get rid of the third loop, I just have to convert to list after the loop and then add in the sample_id key.
For the second point, I don't think I can guarantee that the OrderedDicts will be made with the feature_ids in the same order.  I don't think I can make any assumptions about the order that the vizserver will return the data, so I'm assuming that the loop could encounter any sample_id/feature_id pair at any location in the list.  I could have a loop go through each ordered dict and add the missing feature_id keys with None values, and then sort each OrderedDict in the list.  We basically get this for free in the DictWriter by setting `restval = 0` though.  I could alternatively set `restval = None` there.  I think the decision is basically whether we add in Nones and sort the list of dicts ourselves, or we just pass the column names to the output handler and let DictWriter do the sorting and missing data infill.